topi/include/topi/nn/pooling.h

```diff
@@ -231,6 +231,120 @@ inline Tensor pool(const Tensor& x,
                    count_include_pad);
 }
 
+
+inline Expr start_index(const Var& out_index,
+                        const Expr& odim,
+                        const Expr& idim) {
+  return out_index * idim / odim;
+}
+
+inline Expr end_index(const Var& out_index,
+                      const Expr& odim,
+                      const Expr& idim) {
+  Expr tmp = (out_index + 1) * idim / odim;
+  return tvm::ir::Select::make((out_index + 1) * idim % odim == 0,
+                               tmp, tmp + 1);
+}
+
+/*!
+* \brief Perform adaptive pooling on height and width dimension of data.
+*
+* \param x The input tensor
+* \param output_size Vector of two ints: {output_height, output_width}
+* \param pool_type The type of pooling operator
+* \param height_axis index of the height dimension
+* \param width_axis index of the width dimension
+*
+* \return The output tensor in same layout order
+*/
+inline Tensor adaptive_pool_impl(const Tensor& x,
+                                 const Array<Expr>& output_size,
+                                 PoolType pool_type,
+                                 const size_t height_axis,
+                                 const size_t width_axis) {
+  CHECK_EQ(output_size.size(), 2) << "Pooling kernel_size must have 2 elements";
+
+  auto height = x->shape[height_axis];
+  auto width = x->shape[width_axis];
+
+  auto out_height = output_size[0];
+  auto out_width = output_size[1];
+  Array<Expr> out_shape = x->shape;
+  out_shape.Set(height_axis, out_height);
+  out_shape.Set(width_axis, out_width);
+
+  if (pool_type == kMaxPool) {
+    return tvm::compute(out_shape, [&](const Array<Var>& output) {
+      Array<Expr> indices;
+      for (const Var& var : output) indices.push_back(var);
+      auto i_start_h = start_index(output[height_axis], out_height, height);
+      auto i_end_h = end_index(output[height_axis], out_height, height);
+      auto i_start_w = start_index(output[width_axis], out_width, width);
+      auto i_end_w = end_index(output[width_axis], out_width, width);
+      auto dheight = tvm::reduce_axis(Range(0, i_end_h - i_start_h), "rv1");
```

Comment:
Out of curiosity, is naming the IterVar rv1 necessary?

Reply:
This is optional.